File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        available = set({})

        left = 0
        maxLen = 0

        for right in range(0, len(s)):
            if s[right] in available:
                while left < right and s[left] != s[right]:
                    available.remove(s[left])
                    left += 1

                available.remove(s[left])
                left += 1
            maxLen = max(maxLen, right - left + 1)
            available.add(s[right])
            right += 1

        return maxLen        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int right = 0;
        int n = s.length();
        boolean[] freq = new boolean[128];
        int maxLen = 0;

        while(right < n) {
            char ch = s.charAt(right);

            while(freq[ch]) {
                freq[s.charAt(left)] = false;
                left++;
            }

            freq[ch] = true;

            maxLen = Math.max(maxLen, right - left + 1);
            right++;
        }

        return maxLen;
    }
}